text/0008-ingest-gem-content.md

```diff
@@ -0,0 +1,332 @@
+The files are always accessed directly from block storage through a cache and the manifest is always available before hand.
+
+In another situation, this transparent proxy could make sense.
+It allows the system to store less data and the costs scale exactly with the demand for the feature.
+For rubygems.org, which is does not have a full time staff to maintain a more complex solution, we believe a trade-off towards storage cost and away from operational complexity is warranted.
+
+#### AWS Lambda?
+
+We considered running these on AWS Lambda, using a S3 trigger event when a `.gem` was uploaded.
+The Lambda solution optimizes for higher throughput, but adds costs and complexity.
+For example, while the code in the Lambda function may be similar, it must maintain its own connection to the rubygems.org database in order to save data (whether directly or via an API provided by rubygems.org).
+Testing lambda functions requires more steps and is more involved than testing Rails background jobs.
+The production and staging infrastructure also becomes more complicated, increasing the number of repositories and systems involved in running rubygems.org.
+
+### SHA256 Checksum File Storage
+
+Given the immense size of the uncompressed rubygems.org gem database, it benefits us to make optimizations.
+
+During our research, we were inspired by the solution arrived at by npmjs.com.
+Both the manifest structure and file storage approach is influenced by npmjs.com's design.
+
+When returning files, we serve a manifest that links pathnames to checksums.
+The checksums serve a dual purpose on both npmjs.com and in our solution.
+The first purpose is the common solution of verifiability of file contents.
+The second purpose, which I find clever, is to deduplicate files within a package.
+
+Not every gem version changes every file.
+Most of the files in a gem stay the same between two consecutive releases.
+In our testing, for example, the entire history of rake is about 4 times smaller when deduplicated by matching checksums.
+We therefore propose a solution that uses the checksum of a file as the filename.
+It both increases the likelihood of a cache hit on any given request and reduces the number of files and storage size of almost every gem.
+
+#### Global Checksums
+
+For a time we considered applying this checksum approach globally.
+The apparent benefit would be in deduplicating files across gems.
+Files like LICENSE and other boilerplate could be deduplicated across the entire gem repository.
+Some gems vendor other gems to reduce dependencies (e.g. bundler).
+It seems like there could be enough benefit to warrant this solution.
+
+However, we decided against it for a few reasons, the primary being yanking gems.
+If we had to consider every file as a possible orphan, we would have to search the entire file table without constraints for a matching checksum before deleting a file.
```

Comment:
i dont think this holds if we store each file in its own row

Reply:
Responded [above](https://github.com/rubygems/rfcs/pull/44#discussion_r1085998314) to keep the discussion in one place.